sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -133,39 +133,90 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
+
+      declareAfterPrefix match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable].name.
+              asInstanceOf[UnresolvedIdentifier].nameParts.last,
+            parsedPlan.origin.line.get.toString)
+        case _ =>
+      }
+
+    } else {
+      val declare = compoundStatements
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+
+      declare match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable].name.
+              asInstanceOf[UnresolvedIdentifier].nameParts.last,
+            parsedPlan.origin.line.get.toString)
+        case _ =>
+      }
+    }
+
     CompoundBody(buff.toSeq, label)
   }
 
-  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody = {
-    val beginLabelCtx = Option(ctx.beginLabel())
-    val endLabelCtx = Option(ctx.endLabel())
+  private def generateLabelText(
+                                 beginLabelCtx: Option[BeginLabelContext],
+                                 endLabelCtx: Option[EndLabelContext]) = {
 
     (beginLabelCtx, endLabelCtx) match {
       case (Some(bl: BeginLabelContext), Some(el: EndLabelContext))
         if bl.multipartIdentifier().getText.nonEmpty &&
           bl.multipartIdentifier().getText.toLowerCase(Locale.ROOT) !=
-              el.multipartIdentifier().getText.toLowerCase(Locale.ROOT) =>
+            el.multipartIdentifier().getText.toLowerCase(Locale.ROOT) =>
         throw SqlScriptingErrors.labelsMismatch(
           bl.multipartIdentifier().getText, el.multipartIdentifier().getText)
       case (None, Some(el: EndLabelContext)) =>
         throw SqlScriptingErrors.endLabelWithoutBeginLabel(el.multipartIdentifier().getText)
       case _ =>
     }
 
-    val labelText = beginLabelCtx.
-      map(_.multipartIdentifier().getText).getOrElse(java.util.UUID.randomUUID.toString).
-      toLowerCase(Locale.ROOT)
-    visitCompoundBodyImpl(ctx.compoundBody(), Some(labelText))
+    beginLabelCtx
+      .map(_.multipartIdentifier().getText).getOrElse(java.util.UUID.randomUUID.toString)
+      .toLowerCase(Locale.ROOT)
+  }
+
+  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody = {
+    val labelText = generateLabelText(Option(ctx.beginLabel()), Option(ctx.endLabel()))
+
+    visitCompoundBodyImpl(ctx.compoundBody(), Some(labelText), allowVarDeclare = true)
+  }
+
+  override def visitWhileStatement(ctx: WhileStatementContext): WhileStatement = {
+    val labelText = generateLabelText(Option(ctx.beginLabel()), Option(ctx.endLabel()))
+    val boolExpr = ctx.booleanExpression()
+    WhileStatement(
```

Comment:
use named parameters when having such complicated constructs, so it's a bit cleaner what each parameter means. also, maybe move `SingleStatement(` to the next line, so each parameter has a new level of identation.